### PR TITLE
fix(zone.js): async-test should only call done once

### PR DESCRIPTION
**Pull request by JiaLiPassion**
`AsyncTestZoneSpec` triggers jasmine `done()` function multiple times
and causes warning

```
An asynchronous function called its 'done' callback more than once. This is a bug in the spec, beforeAll, beforeEach, afterAll, or afterEach function in question. This will be treated as an error in a future version. See<https://jasmine.github.io/tutorials/upgrading_to_Jasmine_4.0#deprecations-due-to-calling-done-multiple-times> for more information
```

The reproduce case will be running some `Zone.run()` inside
`waitForAsync()`.

```
it('multiple done', waitForAsync(() => {
  Zone.current.run(() => {});
  Zone.current.run(() => {});
}));
```

The reason the `done()` is called in the `onInvoke()` hook is to handle
the case that the testBody is totally sync, but we should only do this
check for the entry function not for all `Zone.run()` scenario.

Another issue is if we have nested zone inside `waitForAsync()`, the 
`onHasTask()` hook will be triggered multiple times and cause the `done()` 
callback to be triggered multiple times, so we should only trigger it 
when the `zone` is `AsyncTestZone`.

This fix will also fix a lot of flaky errors caused by `done()` is being 
invoked too earlier.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

related to this PR https://github.com/angular/angular/pull/44923

### CODE REVIEW

File: packages/zone.js/lib/zone-spec/async-test.ts [hunk 1]
```diff
@@ -108,13 +109,15 @@ class AsyncTestZoneSpec implements ZoneSpec {
...
+    if (!this.entryFunction) {
+      this.entryFunction = delegate;
+    }
     try {
       this._isSync = true;
       return parentZoneDelegate.invoke(targetZone, delegate, applyThis, applyArgs, source);
     } finally {
       const afterTaskCounts: any = (parentZoneDelegate as any)._taskCounts;
-      if (this._isSync) {
+      if (this._isSync && this.entryFunction === delegate) {
```
Comment (crisbeto) [inline packages/zone.js/lib/zone-spec/async-test.ts:120]: Should we have a comment about when the `entryFunction` could be different?

**Review by crisbeto — APPROVED** [review]
LGTM

File: packages/zone.js/lib/zone-spec/async-test.ts (hunk 1 above)
Comment (JiaLiPassion) [reply packages/zone.js/lib/zone-spec/async-test.ts:120]: Got it, I added comment here.

**Comment by crisbeto** [comment]
Presubmit passed. Marking as merge-ready as discussed.

**Comment by alxhub** [comment]
This PR was merged into the repository by commit d5565ccdb4573a47eb329b09c6852c1ae39672a6.

**Comment by coyoteecd** [comment]
@JiaLiPassion @alxhub is it possible to include this fix in the v13 branch? We're stuck to an older jasmine version because of it, and getting to v14 will probably take a while until the entire ecosystem catches up.

**Comment by JiaLiPassion** [comment]
@coyoteecd, this fix need to be released with new version of `zone.js`, so please wait for the new release of the `zone.js`.

**Pull request closed without merging**